src/NHibernate/Hql/Ast/ANTLR/Tree/DotNode.cs

```diff
@@ -426,13 +429,14 @@ private void DereferenceEntity(EntityType entityType, bool implicitJoin, string
 			else
 			{
 				joinIsNeeded = generateJoin || (Walker.IsInSelect && !Walker.IsInCase) || (Walker.IsInFrom && !Walker.IsComparativeExpressionClause)
-				               || comparisonWithNullableEntity;
+				               || comparisonWithNullableEntityOrThroughPropertyRef;
 			}
 
 			if ( joinIsNeeded )
 			{
-				DereferenceEntityJoin(classAlias, entityType, implicitJoin, parent, comparisonWithNullableEntity);
-				if (comparisonWithNullableEntity)
+				var forceLeftJoin = comparisonWithNullableEntityOrThroughPropertyRef && !IsCorrelatedSubselect;
```

Comment:
CorrelatedSubselect forces a theta-style join, which is not compatible with left joins. That was the cause for #3306 fixed by #3312 by ceasing to generate joins for nullable entities when in a subselect. It looks like a crutch actually. And it remains one.